src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,59 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide: 
+            // "The overall performance is poorer than if the function is called with a zero output buffer"
+            // "A recommended size for the temporary buffer is four times the length of the longest string processed"
+            int recommendedSize = int.MaxValue / 4 > source.Length ? source.Length * 4 : source.Length;
```

Comment:
It is not possible to allocate `new byte[int.MaxValue]` array. The cut-off here should be less than to not just avoid integer overflows, but also make the very large strings work as much as possible.

Reply:
should I just use `Array.MaxByteArrayLength` instead of `int.MaxValue`?

Reply:
I think it should be a lot less - like 1MB. When dealing with very large strings, it is important to allocate only as much as you need to make things work and not run out of memory, or hit other limits. The performance is secondary because of it will be bad anyway.

For example, ArrayPool is won't pool the very large buffers for you either for the same reason. It will always allocate a fresh one for you.

Reply:
What would happen if for anything over 1M we made the size 0 and didn't allocate anything in managed?

Reply:
> What would happen if for anything over 1M we made the size 0 and didn't allocate anything in managed?

The `ucol_getSortKey` method switches then to `preflighting` mode, which is more expensive but returns the actual sort key. After that, we need to allocate the memory of size equal to result returned by `ucol_getSortKey` and pass it to `ucol_getSortKey` again.
This is what we had so far.  If I do that we get the "fast path" for small strings and old, slow for bigger ones.

Reply:
> old, slow for bigger ones

The problem I have is we end up allocating a gigantic LOH array just for doing a single string.GetHashCode. That's unexpected, and has process-wide perf implications beyond the throughout measured by this benchmark.

Reply:
I have changed the code to not try to guess for > 1MB as recommended by @jkotas .

As soon as my VM finishes building CoreCLR I am going to re-run the benchmarks and share the results.

I was also very much surprised when I found out that getting a hash code might allocate byte array.

Reply:
>  anything over 1M we made the size 0 

I think that it is a good idea to make the initial guessed size 0 for the large strings.

> allocating a gigantic LOH array just for doing ...

We have the same problem in many other places that use ArrayPool. If it becomes a problem, I think that this should be fixed in ArrayPool - by having it more integrated with the GC.

Or did you have in mind a streaming hash implementation for large string?

Reply:
> If it becomes a problem, I think that this should be fixed in ArrayPool - by having it more integrated with the GC.

I agree.  At the same time, I think there are alternatives to consider. For example, rather than allocating the array on the LOH and immediately pinning it just to pass it to native, after which we throw away the array, we could instead just allocate the array in native code, either via Marshal.* or in the native shim.  It's still an allocation, but with less impact since it won't impact GC.

Reply:
> we could instead just allocate the array in native code, either via Marshal.

The side effect of this could be unmanaged heap segmentation (GC takes care of LOH segmentation, I am not sure about it when it comes to Marshal.*).

Reply:
>  It's still an allocation, but with less impact since it won't impact GC.

For my education, how is a native allocation less impactful than a LOH allocation? Either way there's some kind of heap manager that finds a chunk, the memory is not moved (as far as I know), etc.

Reply:
A regular LOH allocation requires Gen2 GC to run to reclaim/reuse. It may take a while until that happens.

Native heap allocation can be reclaimed/reused right away as soon as the memory is freed.

ArrayPool allocation can be in theory reclaimed/reused right away as well, but it is not how ArrayPool works today: https://github.com/dotnet/coreclr/blob/4ca032d49b0718eb40f2105f94db2b243cf94957/src/System.Private.CoreLib/shared/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs#L34

Reply:
> ArrayPool allocation can be in theory reclaimed/reused right away as well, but it is not how ArrayPool works today:

Can you clarify, I thought that was the purpose of the ArrayPool - to hold the memory so it's not reclaimed immediately. Do you mean it could have some aging policy?

Reply:
ArrayPool is caching blocks up to certain size limit (<=16MB). It gives up above that limit and just forwards the calls to the GC. My comment was referring to what happens for the blocks above this limit.

Reply:
The various limits were put in place before the ArrayPool had some basic trimming support added in.  Now that that's there, we should reconsider a bunch of these, e.g. whether we should allow the pool to store more than X arrays per bucket, whether we should remove the upper bound on bucket size entirely, etc.

Reply:
Ah yes of course the Trimming that @JeremyKuhne did. Thanks.